source/common/upstream/health_checker_impl.cc

```diff
@@ -431,6 +434,7 @@ void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onInterval() {
 }
 
 void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onTimeout() {
+  expect_reset_ = true;
```

Comment:
I don't think this is right? Don't we want the timeout to result in a failure? Can you add a test that covers this?

Reply:
I *think* this is right, at least I'm pretty sure it preserves behavior.

I added a unit test, `DoubleTimeout` that verifies if there are two timeouts we mark the host as unhealthy; the old `Timeout` test did Timeout-then-Remote-Close which might have skipped some logic.

Here's what I think happened before I showed up in the timeout case:

1. onTimeout() fires.
```
void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onTimeout() {
  host_->setActiveHealthFailureType(Host::ActiveHealthFailureType::TIMEOUT);
  client_->close(Network::ConnectionCloseType::NoFlush);
}
```

2. `setActiveHealthFailureType` takes care of handling that we had an active health failure.

3. The `client_->close()` triggers `onEvent(Network::ConnectionEvent::LocalClose)`.

4. The old code skips calling `handleFailure` (since it's a `LocalClose`), but does `std::move(client_)` to the `deferredDelete` list.

```
void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onEvent(Network::ConnectionEvent event) {
  if (event == Network::ConnectionEvent::RemoteClose) {
    handleFailure(envoy::data::core::v2alpha::HealthCheckFailureType::NETWORK);
  }

  if (event == Network::ConnectionEvent::RemoteClose ||
      event == Network::ConnectionEvent::LocalClose) {
    parent_.dispatcher_.deferredDelete(std::move(client_));
  }

  if (event == Network::ConnectionEvent::Connected && parent_.receive_bytes_.empty()) {
    // In this case we are just testing that we can connect, so immediately succeed. Also, since
    // we are just doing a connection test, close the connection.
    // NOTE(mattklein123): I've seen cases where the kernel will report a successful connection, and
    // then proceed to fail subsequent calls (so the connection did not actually succeed). I'm not
    // sure what situations cause this. If this turns into a problem, we may need to introduce a
    // timer and see if the connection stays alive for some period of time while waiting to read.
    // (Though we may never get a FIN and won't know until if/when we try to write). In short, this
    // may need to get more complicated but we can start here.
    // TODO(mattklein123): If we had a way on the connection interface to do an immediate read (vs.
    // evented), that would be a good check to run here to make sure it returns the equivalent of
    // EAGAIN. Need to think through how that would look from an interface perspective.
    // TODO(mattklein123): In the case that a user configured bytes to write, they will not be
    // be written, since we currently have no way to know if the bytes actually get written via
    // the connection interface. We might want to figure out how to handle this better later.
    client_->close(Network::ConnectionCloseType::NoFlush);
    handleSuccess(false);
  }
}
```

5. Sometime later, `onInterval()` fires and we re-start health checking (confirmed, I believe, by unit tests that check that `enableTimer()` was called on `interval_timer_`, including the new one `DoubleTimeout`).

I could definitely be missing something or misreading here.  Or I could be preserving incorrect behavior.

Reply:
I think this was a previous bug? Pretty sure a timeout should cause a failure to be accounted for?

Reply:
I'm not sure, but I can try to do `expect_reset_ = false`.  However, the Http and Grpc checkers do it the same way I have here (`expect_reset_ = true`)

https://github.com/envoyproxy/envoy/blob/4aec400e9c61c7ee03ea5e6cb5b045b6a7b5abe9/source/common/upstream/health_checker_impl.cc#L285-L296

https://github.com/envoyproxy/envoy/blob/4aec400e9c61c7ee03ea5e6cb5b045b6a7b5abe9/source/common/upstream/health_checker_impl.cc#L671-L676

and it looks like this change will break some existing unit tests as well.  I guess I thought the `setActiveHealthFailureType()` covered what we want (marking the node unhealthy after two failures).

Reply:
I need to look at the code, something seems off here. Thanks for the detailed notes. I will get back to you in a bit.

Reply:
No problem, and if you decide now is the time to address bigger things, I'm happy to do what I can to help, just didn't want to go wayyy off in the weeds :-)

Reply:
I just looked at the currently code, and unless I'm missing something I think this is a bug that was pre-existing that we should fix. I'm fairly certain that if we time out we should be ultimately immediately calling `handleFailure()`. Right? I don't think the current code does that. The reason that HTTP and gRPC work OK is that in the timeout case they will reset the stream, and that will call `handleFailure()`. 

I would add a TCP test where it only takes a single failure to cause a host to become HC failed, and verify that a single timeout does that and fires callbacks. This code will set the active failure type, but it will not cause callbacks to be invoked I think.

/wait

Reply:
OK, I added a unit test for a single HC timeout (check out `TcpHealthCheckerImplTest.Timeout`), verifying that `logEjectUnhealthy` and `logUnhealthy` are called and that status is set to Unhealthy.  Is this what you're talking about or are there other callbacks you're concerned about?  (None jumped out at me from other unit tests).

Reply:
Also I'm not sure that HTTP or gRPC work differently; I guess I'd say to my eyes they're behaving very similarly to the way this PR is currently treating TCP HC timeouts.

HTTP calls `client_->close()` immediately after setting `expect_reset_ = true`:

https://github.com/envoyproxy/envoy/blob/bc4f1d7396a4634f81e7aecc2e18543afbdcc15b/source/common/upstream/health_checker_impl.cc#L291-L294

and the HTTP `onEvent()` doesn't call `handleFailure()` on a local close either:(https://github.com/envoyproxy/envoy/blob/bc4f1d7396a4634f81e7aecc2e18543afbdcc15b/source/common/upstream/health_checker_impl.cc#L178-L186

and looks like gRPC sets `expected_reset_ = true`: https://github.com/envoyproxy/envoy/blob/bc4f1d7396a4634f81e7aecc2e18543afbdcc15b/source/common/upstream/health_checker_impl.cc#L671

and then uses this specifically to bypass calling `handleFailure()`:

https://github.com/envoyproxy/envoy/blob/bc4f1d7396a4634f81e7aecc2e18543afbdcc15b/source/common/upstream/health_checker_impl.cc#L597-L601

Reply:
Yeah I think this is all broken, which is pretty surprising to me. @snowp can you PTAL at this to make sure I'm not crazy?

Reply:
Yeah I agree that this seems broken for HTTP/gRPC as well. We should probably be explicitly calling handleFailure or let it flow through to the reset handler

Reply:
Thanks @snowp. @andrewjjenkins can you fix those cases also and make sure we have tests that the callbacks are called after a single timeout? Thank you so much!

Reply:
OK, I pushed additional unit tests for HTTP and gRPC so I think I'm confirming that we are marking unhealthy and calling callbacks after one timeout:

- HTTP: `HttpHealthCheckerImplTest.Timeout`
- gRPC: `GrpcHealthCheckerImplTest.Timeout`
- TCP: `TcpHealthCheckerImplTest.Timeout`

So just double-checking, AFAICT we are calling the callbacks in all cases. I'm trying to confirm we believe the code as-is in this PR is giving us the correct result (mark unhealthy, call callbacks, even after one timeout).

It sounds like the shortcoming y'all are seeing is related to not calling `handleFailure()`, but it looks to me like that's done by the base impl:
https://github.com/envoyproxy/envoy/blob/6178aeac7b446e3d5121bcc06ae85a6962a72ebb/source/common/upstream/health_checker_base_impl.cc#L361-L364

So I guess I'm not sure what to change - are we trying to pull the call to `handleFailure()` out of `onTimeoutBase()` or something else?

Reply:
Ah OK, sorry, that's what @snowp and I were missing from code inspection. There is no bug here because of calling handleFailure() in the base class timeout function. So I think there is nothing to fix. Let's keep the extra tests though unless they are duplicates of something else. Given that, is this PR ready to review again? Sorry for the runaround!

Reply:
No problem, always a pleasant surprise when things are less broken than you had guessed :-)

Yes, I think it's ready-to-review.  I am OK removing some of the single-timeout tests if we think there's not enough value to justify carrying them along, let me know what you think after looking at the code.